src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs

```diff
@@ -222,13 +224,19 @@ internal static InnerSafeCloseSocket CreateWSASocket(AddressFamily addressFamily
 
             internal static InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle, byte[] socketAddress, ref int socketAddressSize)
             {
-                InnerSafeCloseSocket result = Interop.Winsock.accept(socketHandle.DangerousGetHandle(), socketAddress, ref socketAddressSize);
+                InnerSafeCloseSocket result = Interop.Winsock.accept(socketHandle, socketAddress, ref socketAddressSize);
                 if (result.IsInvalid)
                 {
                     result.SetHandleAsInvalid();
                 }
                 return result;
             }
+
+            internal unsafe bool TryUnblockSocket()
+            {
+                // Try to cancel all pending IO.
+                return Interop.Kernel32.CancelIoEx(this, null);
```

Comment:
Remind me what makes the whole operation thread-safe with regards to race conditions that would cause an additional operation to be created after this call to CancelIoEx (in which case that additional operation wouldn't be canceled by this CancelIoEx and would remain blocked)?  There's something somewhere that ensures that by the time we make this call, no additional operations can be created?

Reply:
Starting a new operation will throw ODE from the Socket class.